apps/zipsync/src/pack.ts

```diff
@@ -0,0 +1,419 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as fs from 'node:fs';
+import * as path from 'node:path';
+import * as crypto from 'node:crypto';
+import * as zlib from 'node:zlib';
+
+import type { ITerminal } from '@rushstack/terminal/lib/ITerminal';
+
+import { crc32Builder } from './crc32';
+import { DISPOSE_SYMBOL, getDisposableFileHandle, type IDisposableFileHandle } from './fs';
+import { type IIncrementalZlib, type IncrementalZlibMode, createIncrementalZlib } from './compress';
+import { markStart, markEnd, getDuration, emitSummary, formatDuration } from './perf';
+import {
+  writeLocalFileHeader,
+  writeDataDescriptor,
+  writeCentralDirectoryHeader,
+  writeEndOfCentralDirectory,
+  ZSTD_COMPRESSION,
+  DEFLATE_COMPRESSION,
+  STORE_COMPRESSION,
+  type ZipMetaCompressionMethod,
+  type IFileEntry,
+  dosDateTime
+} from './zipUtils';
+import { calculateSHA1 } from './hash';
+import {
+  type ZipSyncOptionCompression,
+  type IMetadata,
+  type IDirQueueItem,
+  METADATA_VERSION,
+  METADATA_FILENAME
+} from './zipSyncUtils';
+
+/**
+ * File extensions for which additional DEFLATE/ZSTD compression is unlikely to help.
+ * Used by the 'auto' compression heuristic to avoid wasting CPU on data that is already
+ * compressed (images, media, existing archives, fonts, etc.).
+ */
+const LIKELY_COMPRESSED_EXTENSION_REGEX: RegExp =
+  /\.(?:zip|gz|tgz|bz2|xz|7z|rar|jpg|jpeg|png|gif|webp|avif|mp4|m4v|mov|mkv|webm|mp3|ogg|aac|flac|pdf|woff|woff2)$/;
+
+/**
+ * Map zip compression method code -> incremental zlib mode label
+ */
+const zlibPackModes: Record<ZipMetaCompressionMethod, IncrementalZlibMode | undefined> = {
+  [ZSTD_COMPRESSION]: 'zstd-compress',
+  [DEFLATE_COMPRESSION]: 'deflate',
+  [STORE_COMPRESSION]: undefined
+} as const;
+
+/**
+ * Public facing CLI option -> actual zip method used for a file we decide to compress.
+ */
+const zipSyncCompressionOptions: Record<ZipSyncOptionCompression, ZipMetaCompressionMethod> = {
+  store: STORE_COMPRESSION,
+  deflate: DEFLATE_COMPRESSION,
+  zstd: ZSTD_COMPRESSION,
+  auto: DEFLATE_COMPRESSION
+} as const;
+
+/**
+ * @public
+ * Options for zipsync
+ */
+export interface IZipSyncPackOptions {
+  /**
+   * @rushstack/terminal compatible terminal for logging
+   */
+  terminal: ITerminal;
+  /**
+   * Zip file path
+   */
+  archivePath: string;
+  /**
+   * Target directories to pack or unpack (depending on mode)
+   */
+  targetDirectories: ReadonlyArray<string>;
```

Comment:
These are expected to be relative to `baseDir`, correct?

Reply:
yes. updated comment